# format specifier

# Format Specifiers in Python

A format specifier is used to control how a value is displayed when we print or format a string.

Format specifiers are commonly used with f-strings and the .format() method.

# For example:

In [1]:
name = "Suraj"
age = 35

print(f"My name is {name} and I am {age} years old.")

My name is Suraj and I am 35 years old.


When you want to control the number of decimal places, alignment, percentage, commas, etc., format specifiers become very useful.

# 1. Basic Syntax

* With an f-string:

* f"{value:format_specifier}"

# Example:

In [2]:
price = 125.6789

print(f"{price:.2f}")

125.68


# Here:

* price → value
* : → starts the format specification
* .2f → display 2 decimal places

# Important Format Specifiers

| Specifier | Meaning             | Example       |
| --------- | ------------------- | ------------- |
| `d`       | Integer             | `{num:d}`     |
| `f`       | Floating-point      | `{num:f}`     |
| `.2f`     | 2 decimal places    | `{num:.2f}`   |
| `%`       | Percentage          | `{value:.2%}` |
| `,`       | Thousands separator | `{salary:,}`  |
| `e`       | Scientific notation | `{num:e}`     |
| `s`       | String              | `{name:s}`    |
| `>`       | Right alignment     | `{name:>10}`  |
| `<`       | Left alignment      | `{name:<10}`  |
| `^`       | Center alignment    | `{name:^10}`  |
| `0`       | Zero padding        | `{num:05d}`   |


# Example 1: Decimal Places

* This is one of the most common uses.

In [3]:
price = 125.6789

print(f"Original price: {price}")
print(f"Price: {price:.2f}")
print(f"Price: {price:.3f}")

Original price: 125.6789
Price: 125.68
Price: 125.679


# Example 2: Percentage Formatting

* Suppose a model has an accuracy of 0.9567.

In [4]:
accuracy = 0.9567

print(f"Model Accuracy: {accuracy:.2%}")

Model Accuracy: 95.67%


# Python converts:

0.9567 → 95.67%

# Data Science Example


In [5]:
accuracy = 0.9235
precision = 0.8912
recall = 0.8756

print(f"Accuracy : {accuracy:.2%}")
print(f"Precision: {precision:.2%}")
print(f"Recall   : {recall:.2%}")

Accuracy : 92.35%
Precision: 89.12%
Recall   : 87.56%


# Example 3: Thousands Separator

* Suppose you have a large salary or revenue value.

In [6]:
salary = 1250000

print(f"Salary: {salary:,}")

Salary: 1,250,000


# Another example

In [7]:
revenue = 125678900

print(f"Revenue: ₹{revenue:,}")

Revenue: ₹125,678,900


# Example 4: Alignment

* Format specifiers can control how text is positioned.

* Left Alignment

In [9]:
name = "Python"

print(f"{name:<15}")# The text is aligned to the left within 15 spaces.

Python         


# Center Alignment

In [ ]:

name = "Python"

print(f"{name:^15}")



# Example 5: Zero Padding

* Zero padding is useful when you want a fixed number of digits.

In [10]:
number = 25

print(f"{number:05d}")

00025


# means:

* 0 → fill with zeros
* 5 → total width of 5 characters
* d → integer

In [11]:
employee_id = 42

print(f"EMP{employee_id:05d}")

EMP00042


# Example 6: Scientific Notation

* Scientific notation is useful for very large or very small numbers.

In [12]:
number = 123456789

print(f"{number:e}")

1.234568e+08


# Example 7: Combining Multiple Format Specifiers

* You can combine formatting options.

In [14]:
price = 1234567.89123

print(f"Price: ₹{price:,.2f}")



# Here:

# ,   → thousands separator
# .2  → 2 decimal places
# f   → floating-point number

# This is a very useful format in real-world applications.

Price: ₹1,234,567.89


# Example 8: Creating a Simple Data Report

* You can combine several format specifiers to create a clean report.

In [16]:
name = "Suraj"
salary = 1250000
accuracy = 0.9456
experience = 5.5

print(f"Name       : {name}")
print(f"Salary     : ₹{salary:,}")
print(f"Experience : {experience:.1f} years")
print(f"Accuracy   : {accuracy:.2%}")


Name       : Suraj
Salary     : ₹1,250,000
Experience : 5.5 years
Accuracy   : 94.56%


# Common Format Specifiers to Remember


| Format | Example     | Result         |
| ------ | ----------- | -------------- |
| `:.2f` | `125.6789`  | `125.68`       |
| `:.3f` | `125.6789`  | `125.679`      |
| `:.2%` | `0.9567`    | `95.67%`       |
| `:,`   | `1250000`   | `1,250,000`    |
| `:05d` | `25`        | `00025`        |
| `:<10` | `"Python"`  | Left aligned   |
| `:>10` | `"Python"`  | Right aligned  |
| `:^10` | `"Python"`  | Center aligned |
| `:.2e` | `123456789` | `1.23e+08`     |


# f-string vs .format()
* Using f-string

In [18]:
price = 125.6789

print(f"Price: ₹{price:.2f}")
# Using .format()
price = 125.6789

print("Price: ₹{:.2f}".format(price))

# Both produce:

# Price: ₹125.68

# For modern Python code, f-strings are generally the easiest and most readable approach.

Price: ₹125.68
Price: ₹125.68


# Interview Answer

Format specifiers in Python are used to control how values are displayed when formatting strings. They are commonly used with f-strings. For example, .2f displays two decimal places, ,.2f adds thousands separators and two decimal places, % formats a value as a percentage, and <, >, and ^ control text alignment.